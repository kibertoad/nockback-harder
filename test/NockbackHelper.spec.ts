import request from 'superagent'
import express from 'express'
import rimraf from 'rimraf'
import { initHelper, runSimpleApp, loadJSON, saveJSON } from './utils/testUtils'

describe('NockbackHelper', () => {
  it('block unmocked external call', async () => {
    const helper = initHelper(__dirname)
    helper.startLockdown()
    expect.hasAssertions()

    await helper.nockBack('microsoft.com-GET.json', async () => {
      try {
        await request.get('www.microsoft.com')
      } catch (err) {
        expect(err.message).toMatch('Disallowed net connect')
      }
    })
  })

  it('replay', async () => {
    const helper = initHelper(__dirname)
    helper.startLockdown()

    await helper.nockBack('google.com-GET.json', async () => {
      const response = await request.get('www.google.com')
      expect(response.status).toBe(200)
      expect(response.text).toMatchSnapshot()
    })
  })

  it('replay existing mock but throw on unexpected one', async () => {
    rimraf.sync(getFixturePath('google.com-GET-append.json'))
    const helper = initHelper(__dirname)
    helper.startRecordingNew()
    await helper.nockBack('google.com-GET-append.json', async () => {
      await request.get('www.google.com')
    })

    expect.assertions(1)
    try {
      await helper.nockBack('google.com-GET-append.json', async () => {
        await request.get('www.google.com')
        await request.get('www.microsoft.com')
      })
    } catch (err) {
      expect(err.message).toMatchSnapshot()
    }
  })

  it('can overwrite existing mocks', async () => {
    const fixturePath = getFixturePath('local-GET-overwrite.json')
    const helper = initHelper(__dirname, false)
    helper.startRecordingOverwrite()
    const server = runSimpleApp()
    const [originalMock] = loadJSON(fixturePath)
    expect(originalMock.path).toEqual('/?dummy=value')

    await helper.nockBack('local-GET-overwrite.json', async () => {
      await request.get('localhost:4000').query({
        dummy2: 'value'
      })
    })

    server.close()

    const [updatedMock] = loadJSON(fixturePath)
    expect(updatedMock.path).toEqual('/?dummy2=value')
    saveJSON([originalMock], fixturePath)
  })

  it('can overwrite existing empty mocks', async () => {
    const fixturePath = getFixturePath('local-GET-empty-overwrite.json')
    const helper = initHelper(__dirname, false)
    helper.startRecordingOverwrite()
    const server = runSimpleApp()

    await helper.nockBack('local-GET-empty-overwrite.json', async () => {
      await request.get('localhost:4000').query({
        dummy2: 'value'
      })
    })

    server.close()

    const [updatedMock] = loadJSON(fixturePath)
    expect(updatedMock.path).toEqual('/?dummy2=value')
    saveJSON([], fixturePath)
  })

  it('does not overwrite existing mocks when doNotOverwrite flag is set', async () => {
    const fixturePath = getFixturePath('local-GET-do-not-overwrite.json')
    const helper = initHelper(__dirname, false)
    helper.startRecordingOverwrite()
    expect.hasAssertions()
    const server = runSimpleApp()
    const [originalMock] = loadJSON(fixturePath)
    expect(originalMock.path).toEqual('/?dummy=value')

    await helper.nockBack('local-GET-do-not-overwrite.json', { doNotOverwrite: true }, async () => {
      try {
        await request.get('localhost:4000').query({
          dummy2: 'value'
        })
      } catch (err) {
        expect(err.message).toMatch('No match for request')
      }
    })

    server.close()

    const [updatedMock] = loadJSON(fixturePath)
    expect(updatedMock.path).toEqual('/?dummy=value')
    saveJSON([originalMock], fixturePath)
  })

  it('bypass local', async () => {
    rimraf.sync(getFixturePath('local-GET.json'))
    const helper = initHelper(__dirname)
    helper.startRecordingNew()
    const server = runSimpleApp()

    await helper.nockBack('local-GET.json', async () => {
      const response = await request.get('localhost:4000')
      expect(response.body).toMatchSnapshot()
    })

    server.close()

    const local = loadJSON(getFixturePath('local-GET.json'))
    expect(local).toEqual([])
  })

  it('do not bypass local if whitelisted', async () => {
    rimraf.sync(getFixturePath('local-GET-whitelist.json'))
    const helper = initHelper(__dirname)
    helper.startRecordingOverwrite()
    const server = runSimpleApp()
    const server2 = runSimpleApp(4001)

    await helper.nockBack(
      'local-GET-whitelist.json',
      { passthroughPortWhitelist: [4000] },
      async () => {
        const response = await request.get('localhost:4000')
        expect(response.body).toMatchSnapshot()

        const response2 = await request.get('localhost:4001')
        expect(response2.body).toMatchSnapshot()
      }
    )

    server.close()
    server2.close()

    const local = loadJSON(getFixturePath('local-GET-whitelist.json'))
    expect(local.length).toEqual(1)
    expect(local[0].scope).toEqual('http://localhost:4000')
    rimraf.sync(getFixturePath('local-GET-whitelist.json'))
  })

  it('bypass local when recording on top of empty mocks', async () => {
    const fixturePath = getFixturePath('local-GET-empty-record.json')
    rimraf.sync(fixturePath)
    saveJSON([], fixturePath)

    const helper = initHelper(__dirname)
    helper.startRecordingNew()
    const server = runSimpleApp()

    await helper.nockBack('local-GET-empty-record.json', async () => {
      const response = await request.get('localhost:4000')
      expect(response.body).toMatchSnapshot()
    })

    server.close()

    const local = loadJSON(getFixturePath('local-GET-empty-record.json'))
    expect(local).toEqual([])
  })

  it('bypassing local works with lockdown', async () => {
    const helper = initHelper(__dirname)
    helper.startLockdown()
    const app = express()
    // @ts-ignore
    app.get('/', (req: express.Response, res: express.Response) => {
      res.status(200).send({ status: 'ok' })
    })
    const server = app.listen(4000)

    await helper.nockBack('local-GET-lockdown.json', async () => {
      const response = await request.get('localhost:4000')
      expect(response.body).toMatchSnapshot()
    })

    server.close()
  })

  it('bypassing local works with lockdown and correctly fails on unexpected external calls', async () => {
    jest.setTimeout(3000000)
    const helper = initHelper(__dirname)
    helper.startLockdown()
    const app = express()
    // @ts-ignore
    app.get('/', (req: express.Response, res: express.Response) => {
      res.status(200).send({ status: 'ok' })
    })
    const server = app.listen(4000)

    try {
      expect.assertions(1)
      await helper.nockBack('local-GET-lockdown.json', async () => {
        await request.get('localhost:4000')
        await request.get('www.microsoft.com')
      })
    } catch (err) {
      expect(err.message).toEqual('Nock: Disallowed net connect for "www.microsoft.com:80/"')
    } finally {
      server.close()
    }
  })
})

function getFixturePath(fixtureName: string) {
  return __dirname + '/nock-fixtures/' + fixtureName
}
