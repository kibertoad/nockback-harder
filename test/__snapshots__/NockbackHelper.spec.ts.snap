// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NockbackHelper replay 1`] = `
"<!doctype html><html itemscope=\\"\\" itemtype=\\"http://schema.org/WebPage\\" lang=\\"nl\\"><head><meta content=\\"text/html; charset=UTF-8\\" http-equiv=\\"Content-Type\\"><meta content=\\"/images/branding/googleg/1x/googleg_standard_color_128dp.png\\" itemprop=\\"image\\"><title>Google</title><script nonce=\\"m4+2JkJHf6FRnSgu4QLE7A==\\">(function(){window.google={kEI:'Z22uXIOvCIbTwQL54buIDQ',kEXPI:'0,1353746,58,50,1907,2423,698,527,731,223,756,819,30,1228,1893,58,349,178,350,439,7,221,21,227,142,2333174,403,329157,1294,12383,4855,32691,15248,867,12163,14324,2197,364,3319,5505,2436,266,5107,575,835,284,2,578,728,2431,1362,4323,4968,773,2248,3275,2622,2,1965,2595,2,3599,669,1050,1808,1397,81,7,491,620,29,2373,6722,1209,1288,2,617,3389,797,101,1119,38,920,269,485,119,1215,1366,1611,2693,43,293,1265,1503,2,631,2562,2,4,2,670,44,3744,560,354,124,2608,632,669,429,41,1089,6,650,16,93,229,254,2,526,810,388,143,724,522,478,672,1014,300,578,29,98,756,98,392,29,78,322,376,2,614,455,87,565,10,168,8,87,209,837,76,153,78,365,367,115,315,20,174,47,510,6,404,601,11,14,10,55,303,215,1086,288,11,479,61,76,898,9,25,283,23,194,5,55,423,394,187,193,67,474,154,145,324,176,17,78,32,263,74,84,262,79,26,2,124,151,24,160,68,39,18,146,219,320,1186,482,73,93,548,1517,95,5945652,2554,95,236,5997164,338,2,2799951,4,248,1324,549,333,444,1,2,80,1,900,576,16,304,1,8,1,2,2132,1,1,1,1,1,414,1,748,141,59,726,3,7,563,1,1679,1,55,1,10,42,3338850',authuser:0,kscs:'c9c918f0_Z22uXIOvCIbTwQL54buIDQ',kGL:'NL'};google.sn='webhp';google.kHL='nl';})();(function(){google.lc=[];google.li=0;google.getEI=function(a){for(var b;a&&(!a.getAttribute||!(b=a.getAttribute(\\"eid\\")));)a=a.parentNode;return b||google.kEI};google.getLEI=function(a){for(var b=null;a&&(!a.getAttribute||!(b=a.getAttribute(\\"leid\\")));)a=a.parentNode;return b};google.https=function(){return\\"https:\\"==window.location.protocol};google.ml=function(){return null};google.time=function(){return(new Date).getTime()};google.log=function(a,b,e,c,g){if(a=google.logUrl(a,b,e,c,g)){b=new Image;var d=google.lc,f=google.li;d[f]=b;b.onerror=b.onload=b.onabort=function(){delete d[f]};google.vel&&google.vel.lu&&google.vel.lu(a);b.src=a;google.li=f+1}};google.logUrl=function(a,b,e,c,g){var d=\\"\\",f=google.ls||\\"\\";e||-1!=b.search(\\"&ei=\\")||(d=\\"&ei=\\"+google.getEI(c),-1==b.search(\\"&lei=\\")&&(c=google.getLEI(c))&&(d+=\\"&lei=\\"+c));c=\\"\\";!e&&google.cshid&&-1==b.search(\\"&cshid=\\")&&\\"slh\\"!=a&&(c=\\"&cshid=\\"+google.cshid);a=e||\\"/\\"+(g||\\"gen_204\\")+\\"?atyp=i&ct=\\"+a+\\"&cad=\\"+b+d+f+\\"&zx=\\"+google.time()+c;/^http:/i.test(a)&&google.https()&&(google.ml(Error(\\"a\\"),!1,{src:a,glmm:1}),a=\\"\\");return a};}).call(this);(function(){google.y={};google.x=function(a,b){if(a)var c=a.id;else{do c=Math.random();while(google.y[c])}google.y[c]=[a,b];return!1};google.lm=[];google.plm=function(a){google.lm.push.apply(google.lm,a)};google.lq=[];google.load=function(a,b,c){google.lq.push([[a],b,c])};google.loadAll=function(a,b){google.lq.push([a,b])};}).call(this);google.f={};var a=window.location,b=a.href.indexOf(\\"#\\");if(0<=b){var c=a.href.substring(b+1);/(^|&)q=/.test(c)&&-1==c.indexOf(\\"#\\")&&a.replace(\\"/search?\\"+c.replace(/(^|&)fp=[^&]*/g,\\"\\")+\\"&cad=h\\")};</script><style>#gbar,#guser{font-size:13px;padding-top:1px !important;}#gbar{height:22px}#guser{padding-bottom:7px !important;text-align:right}.gbh,.gbd{border-top:1px solid #c9d7f1;font-size:1px}.gbh{height:0;position:absolute;top:24px;width:100%}@media all{.gb1{height:22px;margin-right:.5em;vertical-align:top}#gbar{float:left}}a.gb1,a.gb4{text-decoration:underline !important}a.gb1,a.gb4{color:#00c !important}.gbi .gb4{color:#dd8e27 !important}.gbf .gb4{color:#900 !important}
</style><style>body,td,a,p,.h{font-family:arial,sans-serif}body{margin:0;overflow-y:scroll}#gog{padding:3px 8px 0}td{line-height:.8em}.gac_m td{line-height:17px}form{margin-bottom:20px}.h{color:#36c}.q{color:#00c}.ts td{padding:0}.ts{border-collapse:collapse}em{font-weight:bold;font-style:normal}.lst{height:25px;width:496px}.gsfi,.lst{font:18px arial,sans-serif}.gsfs{font:17px arial,sans-serif}.ds{display:inline-box;display:inline-block;margin:3px 0 4px;margin-left:4px}input{font-family:inherit}a.gb1,a.gb2,a.gb3,a.gb4{color:#11c !important}body{background:#fff;color:black}a{color:#11c;text-decoration:none}a:hover,a:active{text-decoration:underline}.fl a{color:#36c}a:visited{color:#551a8b}a.gb1,a.gb4{text-decoration:underline}a.gb3:hover{text-decoration:none}#ghead a.gb2:hover{color:#fff !important}.sblc{padding-top:5px}.sblc a{display:block;margin:2px 0;margin-left:13px;font-size:11px}.lsbb{background:#eee;border:solid 1px;border-color:#ccc #999 #999 #ccc;height:30px}.lsbb{display:block}.ftl,#fll a{display:inline-block;margin:0 12px}.lsb{background:url(/images/nav_logo229.png) 0 -261px repeat-x;border:none;color:#000;cursor:pointer;height:30px;margin:0;outline:0;font:15px arial,sans-serif;vertical-align:top}.lsb:active{background:#ccc}.lst:focus{outline:none}</style><script nonce=\\"m4+2JkJHf6FRnSgu4QLE7A==\\"></script></head><body bgcolor=\\"#fff\\"><script nonce=\\"m4+2JkJHf6FRnSgu4QLE7A==\\">(function(){var src='/images/nav_logo229.png';var iesg=false;document.body.onload = function(){window.n && window.n();if (document.images){new Image().src=src;}
if (!iesg){document.f&&document.f.q.focus();document.gbqf&&document.gbqf.q.focus();}
}
})();</script><div id=\\"mngb\\"> <div id=gbar><nobr><b class=gb1>Zoeken</b> <a class=gb1 href=\\"http://www.google.nl/imghp?hl=nl&tab=wi\\">Afbeeldingen</a> <a class=gb1 href=\\"http://maps.google.nl/maps?hl=nl&tab=wl\\">Maps</a> <a class=gb1 href=\\"https://play.google.com/?hl=nl&tab=w8\\">Play</a> <a class=gb1 href=\\"http://www.youtube.com/?gl=NL&tab=w1\\">YouTube</a> <a class=gb1 href=\\"http://news.google.nl/nwshp?hl=nl&tab=wn\\">Nieuws</a> <a class=gb1 href=\\"https://mail.google.com/mail/?tab=wm\\">Gmail</a> <a class=gb1 href=\\"https://drive.google.com/?tab=wo\\">Drive</a> <a class=gb1 style=\\"text-decoration:none\\" href=\\"https://www.google.nl/intl/nl/about/products?tab=wh\\"><u>Meer</u> &raquo;</a></nobr></div><div id=guser width=100%><nobr><span id=gbn class=gbi></span><span id=gbf class=gbf></span><span id=gbe></span><a href=\\"http://www.google.nl/history/optout?hl=nl\\" class=gb4>Webgeschiedenis</a> | <a  href=\\"/preferences?hl=nl\\" class=gb4>Instellingen</a> | <a target=_top id=gb_70 href=\\"https://accounts.google.com/ServiceLogin?hl=nl&passive=true&continue=http://www.google.com/\\" class=gb4>Inloggen</a></nobr></div><div class=gbh style=left:0></div><div class=gbh style=right:0></div> </div><center><br clear=\\"all\\" id=\\"lgpd\\"><div id=\\"lga\\"><img alt=\\"Google\\" height=\\"92\\" src=\\"/images/branding/googlelogo/1x/googlelogo_white_background_color_272x92dp.png\\" style=\\"padding:28px 0 14px\\" width=\\"272\\" id=\\"hplogo\\" onload=\\"window.lol&&lol()\\"><br><br></div><form action=\\"/search\\" name=\\"f\\"><table cellpadding=\\"0\\" cellspacing=\\"0\\"><tr valign=\\"top\\"><td width=\\"25%\\">&nbsp;</td><td align=\\"center\\" nowrap=\\"\\"><input name=\\"ie\\" value=\\"ISO-8859-1\\" type=\\"hidden\\"><input value=\\"nl\\" name=\\"hl\\" type=\\"hidden\\"><input name=\\"source\\" type=\\"hidden\\" value=\\"hp\\"><input name=\\"biw\\" type=\\"hidden\\"><input name=\\"bih\\" type=\\"hidden\\"><div class=\\"ds\\" style=\\"height:32px;margin:4px 0\\"><input style=\\"color:#000;margin:0;padding:5px 8px 0 6px;vertical-align:top\\" autocomplete=\\"off\\" class=\\"lst\\" value=\\"\\" title=\\"Google zoeken\\" maxlength=\\"2048\\" name=\\"q\\" size=\\"57\\"></div><br style=\\"line-height:0\\"><span class=\\"ds\\"><span class=\\"lsbb\\"><input class=\\"lsb\\" value=\\"Google zoeken\\" name=\\"btnG\\" type=\\"submit\\"></span></span><span class=\\"ds\\"><span class=\\"lsbb\\"><input class=\\"lsb\\" value=\\"Ik doe een gok\\" name=\\"btnI\\" onclick=\\"if(this.form.q.value)this.checked=1; else top.location='/doodles/'\\" type=\\"submit\\"></span></span></td><td class=\\"fl sblc\\" align=\\"left\\" nowrap=\\"\\" width=\\"25%\\"><a href=\\"/advanced_search?hl=nl&amp;authuser=0\\">Geavanceerd zoeken</a><a href=\\"/language_tools?hl=nl&amp;authuser=0\\">Google Taalhulpmiddelen</a></td></tr></table><input id=\\"gbv\\" name=\\"gbv\\" type=\\"hidden\\" value=\\"1\\"><script nonce=\\"m4+2JkJHf6FRnSgu4QLE7A==\\">(function(){var a,b=\\"1\\";if(document&&document.getElementById)if(\\"undefined\\"!=typeof XMLHttpRequest)b=\\"2\\";else if(\\"undefined\\"!=typeof ActiveXObject){var c,d,e=[\\"MSXML2.XMLHTTP.6.0\\",\\"MSXML2.XMLHTTP.3.0\\",\\"MSXML2.XMLHTTP\\",\\"Microsoft.XMLHTTP\\"];for(c=0;d=e[c++];)try{new ActiveXObject(d),b=\\"2\\"}catch(h){}}a=b;if(\\"2\\"==a&&-1==location.search.indexOf(\\"&gbv=2\\")){var f=google.gbvu,g=document.getElementById(\\"gbv\\");g&&(g.value=a);f&&window.setTimeout(function(){location.href=f},0)};}).call(this);</script></form><div id=\\"gac_scont\\"></div><div style=\\"font-size:83%;min-height:3.5em\\"><br><div id=\\"prm\\"><style>.szppmdbYutt__middle-slot-promo{font-size:small;margin-bottom:32px}.szppmdbYutt__middle-slot-promo a.ZIeIlb{display:inline-block;text-decoration:none}.szppmdbYutt__middle-slot-promo img{border:none;margin-right:5px;vertical-align:middle}</style><div class=\\"szppmdbYutt__middle-slot-promo\\" data-ved=\\"0ahUKEwjD_uvmyMbhAhWGaVAKHfnwDtEQnIcBCAQ\\"><a class=\\"NKcBbd\\" href=\\"https://www.google.com/url?q=https://about.google/intl/nl_nl/stories/gpsart%3Futm_source%3Dgoogle%26utm_medium%3Dhpp%26utm_campaign%3DNetherlands&amp;source=hpp&amp;id=19011697&amp;ct=3&amp;usg=AFQjCNEyDOPQZ9FcCrML4QEhEf5ZyHr3tA&amp;sa=X&amp;ved=0ahUKEwjD_uvmyMbhAhWGaVAKHfnwDtEQ8IcBCAU\\" rel=\\"nofollow\\">Search On</a><span>: Hoe een Japanse man een huwelijksaanzoek deed met Google Earth</span></div></div><div id=\\"gws-output-pages-elements-homepage_additional_languages__als\\"><style>#gws-output-pages-elements-homepage_additional_languages__als{font-size:small;margin-bottom:24px}#SIvCob{display:inline-block;line-height:28px;}#SIvCob a{padding:0 3px;}.H6sW5{display:inline-block;margin:0 2px;white-space:nowrap}.z4hgWe{display:inline-block;margin:0 2px}</style><div id=\\"SIvCob\\">Google aangeboden in:  <a href=\\"http://www.google.com/setprefs?sig=0_fV8cf1S2m0mVqxoUIwCeZ4G62WI%3D&amp;hl=fy&amp;source=homepage&amp;sa=X&amp;ved=0ahUKEwjD_uvmyMbhAhWGaVAKHfnwDtEQ2ZgBCAc\\">Frysk</a>  </div></div></div><span id=\\"footer\\"><div style=\\"font-size:10pt\\"><div style=\\"margin:19px auto;text-align:center\\" id=\\"fll\\"><a href=\\"/intl/nl/ads/\\">Advertentieprogramma's</a><a href=\\"http://www.google.nl/intl/nl/services/\\">Bedrijfsoplossingen</a><a href=\\"/intl/nl/about.html\\">Alles over Google</a><a href=\\"http://www.google.com/setprefdomain?prefdom=NL&amp;prev=http://www.google.nl/&amp;sig=K_jCGzTpDT3emC4eqO7PTjgRMqXf0%3D\\">Google.nl</a></div></div><p style=\\"color:#767676;font-size:8pt\\">&copy; 2019 - <a href=\\"/intl/nl/policies/privacy/\\">Privacy</a> - <a href=\\"/intl/nl/policies/terms/\\">Voorwaarden</a></p></span></center><script nonce=\\"m4+2JkJHf6FRnSgu4QLE7A==\\">(function(){window.google.cdo={height:0,width:0};(function(){var a=window.innerWidth,b=window.innerHeight;if(!a||!b){var c=window.document,d=\\"CSS1Compat\\"==c.compatMode?c.documentElement:c.body;a=d.clientWidth;b=d.clientHeight}a&&b&&(a!=google.cdo.width||b!=google.cdo.height)&&google.log(\\"\\",\\"\\",\\"/client_204?&atyp=i&biw=\\"+a+\\"&bih=\\"+b+\\"&ei=\\"+google.kEI);}).call(this);})();(function(){var u='/xjs/_/js/k\\\\x3dxjs.hp.en.4w-QH8u0uXQ.O/m\\\\x3dsb_he,d/am\\\\x3dBIOa/rt\\\\x3dj/d\\\\x3d1/rs\\\\x3dACT90oGwJzXJXOIa2xgXKUvNaMLyqzFULA';setTimeout(function(){var a=document.createElement(\\"script\\");a.src=u;google.timers&&google.timers.load&&google.tick&&google.tick(\\"load\\",\\"xjsls\\");document.body.appendChild(a)},0);})();(function(){window.google.xjsu='/xjs/_/js/k\\\\x3dxjs.hp.en.4w-QH8u0uXQ.O/m\\\\x3dsb_he,d/am\\\\x3dBIOa/rt\\\\x3dj/d\\\\x3d1/rs\\\\x3dACT90oGwJzXJXOIa2xgXKUvNaMLyqzFULA';})();function _DumpException(e){throw e;}
(function(){google.spjs=false;})();(function(){var pmc='{\\\\x22Qnk92g\\\\x22:{},\\\\x22RWGcrA\\\\x22:{},\\\\x22U5B21g\\\\x22:{},\\\\x22YFCs/g\\\\x22:{},\\\\x22ZI/YVQ\\\\x22:{},\\\\x22d\\\\x22:{},\\\\x22sb_he\\\\x22:{\\\\x22agen\\\\x22:true,\\\\x22cgen\\\\x22:true,\\\\x22client\\\\x22:\\\\x22heirloom-hp\\\\x22,\\\\x22dh\\\\x22:true,\\\\x22dhqt\\\\x22:true,\\\\x22ds\\\\x22:\\\\x22\\\\x22,\\\\x22ffql\\\\x22:\\\\x22en\\\\x22,\\\\x22fl\\\\x22:true,\\\\x22host\\\\x22:\\\\x22google.com\\\\x22,\\\\x22isbh\\\\x22:28,\\\\x22jsonp\\\\x22:true,\\\\x22msgs\\\\x22:{\\\\x22cibl\\\\x22:\\\\x22Zoekopdracht wissen\\\\x22,\\\\x22dym\\\\x22:\\\\x22Bedoelde je:\\\\x22,\\\\x22lcky\\\\x22:\\\\x22Ik doe een gok\\\\x22,\\\\x22lml\\\\x22:\\\\x22Meer informatie\\\\x22,\\\\x22oskt\\\\x22:\\\\x22Invoertools\\\\x22,\\\\x22psrc\\\\x22:\\\\x22Deze zoekopdracht is verwijderd uit je \\\\\\\\u003Ca href\\\\x3d\\\\\\\\\\\\x22/history\\\\\\\\\\\\x22\\\\\\\\u003EWebgeschiedenis\\\\\\\\u003C/a\\\\\\\\u003E\\\\x22,\\\\x22psrl\\\\x22:\\\\x22Verwijderen\\\\x22,\\\\x22sbit\\\\x22:\\\\x22Zoeken op afbeelding\\\\x22,\\\\x22srch\\\\x22:\\\\x22Google zoeken\\\\x22},\\\\x22ovr\\\\x22:{},\\\\x22pq\\\\x22:\\\\x22\\\\x22,\\\\x22refpd\\\\x22:true,\\\\x22rfs\\\\x22:[],\\\\x22sbpl\\\\x22:24,\\\\x22sbpr\\\\x22:24,\\\\x22scd\\\\x22:10,\\\\x22sce\\\\x22:5,\\\\x22stok\\\\x22:\\\\x224nhRBzmz2PBeEM2mVGRX6XoUGkU\\\\x22,\\\\x22uhde\\\\x22:false}}';google.pmc=JSON.parse(pmc);})();</script>        </body></html>"
`;
